src/fmio.1
.\"
.\" $Id$
.\"
.\" Copyright (c) 2000 - 2002 Vladimir Popov <jumbo@narod.ru>.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\" OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.Dd November 20, 2000
.Dt fmio 1
.Os
.Sh NAME
.Nm fmio
.Nd small utility to manipulate fm radio card
.Sh SYNOPSIS
.Nm fmio
.Op Fl d Ar driver
.Op Fl f Ar freq
.Op Fl i
.Op Fl m
.Op Fl s
.Op Fl v Ar vol
.Op Fl x Ar vol
.Op Fl X Ar vol
.Nm fmio
.Op Fl d Ar driver
.Fl S
.Op Fl c Ar count
.Op Fl l Ar begin
.Op Fl h Ar end
.Nm fmio
.Op Fl d Ar driver
.Fl W Ar frequency
.Nm fmio
.Fl D
.Sh DESCRIPTION
The
.Nm
is a small program to set and change fm radio card parameters.
.Pp
It sets frequency, volume of a card, reports signal strength on
the working frequency and can set a radio card output to mono
provided that the card driver supports it.
.Pp
The
.Nm
has two additional modes - scanning and detection.
In the scan mode
.Nm
scans frequency range with step 10 kHz and outputs measured signal strength.
Frequency range and number of probes for each frequency are configurable.
In the detection mode
.Nm
detects any supported radio card.
Use this mode if you are not sure which card you have or do not know which
port your card uses.
.Pp
Also, it can control the output volume of a sound card through
.Pa /dev/mixer .
.Pp
The options are as follows:
.Bl -tag -width "-m "
.It Fl d Ar driver
Use driver
.Ar driver .
See list of currently known drivers in section
.Sx DRIVERS .
If no driver was specified, default one will be used.
.It Fl f Ar freq
Set fm card frequency
.Pq in MHz .
.It Fl i
Show all available information about a driver and a card.
.It Fl m
Set output of a card to mono.
Works only in conjunction with option
.Fl f .
Note, that not all drivers support this option.
.It Fl s
Show current signal status of a radio card.
.It Fl v Ar volume
Set volume of a card.
Option
.Fl v Ar 0
will mute a card.
Each driver has its own maximal volume.
.It Fl x Ar volume
Set volume of an arbitrary mixer line of a sound card thru
.Pa /dev/mixer .
The
.Ar volume
has following format:
.Pp
.Dl <line name>=[+-]<left channel>,[+-]<right channel>.
.Pp
For the line names see
.Dl under OpenBSD and NetBSD: Pa /usr/include/sys/audioio.h
.Dl under Linux and FreeBSD: Pa /usr/include/sys/soundcard.h
.Pp
Some examples of the option
.Fl x
usage are in the section
.Sx EXAMPLES
below.
.It Fl X Ar volume
Set volume of the sound card master output through
.Pa /dev/mixer .
Format of this option is similar to option
.Fl x ,
except there is no need to specify <line name>.
.It Fl D
Detection mode. All known cards will be probed and results will be printed to
standard output.
Note, this procedure is very slow.
.It Fl S
Scan mode.
If some of the listed below parameters were not specified the FM maximal and
minimal frequencies are used.
Measured data is dumped to standard output.
.It Fl l Ar begin
Lower border of scan range
.Pq in MHz .
If it is less than the FM minimal frequency
.Pq 87.5 MHz ,
it is set to the FM minimal frequency.
.It Fl h Ar end
Upper border of scan range
.Pq in MHz .
If it is greater than the FM maximal frequency
.Pq 108.0 MHz ,
it is set to the FM maximal frequency.
.It Fl c Ar count
Number of probes for each frequency.
The higher this number is the more precise will be results.
Since stereo signal has weight 3, none - weight 0 and mono is in between,
the strongest signal will have value 3 * 
.Ar count .
If not set, each frequency will be probed only once.
.It Fl W Ar frequency
Search mode.
Nearest to
.Ar frequency
station is searched and, if found, tuned to.
The search direction is determined by
.Ar frequency
sign.
If
.Ar frequency
has negative value, the station is searched downward.
If
.Ar frequency
is positive the search is performed upward.
.El
.Sh DRIVERS
.Pp
.Em Aztech/PackardBell FM Radio Card
.Pq ISA
.Ql az
.Dl port 0x350 - az1
.Dl port 0x358 - az2
.Dl Volume - 0 .. 3
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em Brooktree Bt848 Driver
.Ql bktr
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Software search
.Dl Use with Nm bktrctl
.Pp
.Em BMC FM Radio Card
.Pq ISA
.Ql bmc
.Dl port 0x20F - bmc1
.Dl port 0x30F - bmc2
.Dl Volume - 0 .. 15
.Dl Can set mono - yes
.Dl Search not implemented yet
.Pp
.Em BSD Radio Driver
.Ql br
.Dl Volume - 0 .. 255
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em EcoRadio Card
.Pq ISA
.Ql er
.Dl port 0x316 - er1
.Dl port 0x336 - er2
.Dl Volume - 0 .. 3
.Dl Can set mono - no
.Dl Search not implemented yet
.Dl Detection not implemented yet
.Pp
.Em Gemtek FM Radio Card
.Pq ISA
.Ql gti
.Dl port 0x20c - gti1
.Dl port 0x30c - gti2
.Dl port 0x24c - gti3
.Dl port 0x34c - gti4
.Dl port 0x248 - gti5
.Dl Volume - 0 .. 1
.Dl Can set mono - no
.Dl Software search
.Pp
.Em Gemtek FM Radio Card
.Pq PCI
.Ql gtp
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em AIMS Lab Highway Xtreme TV card
.Em with FM tuner Driver
.Ql hx
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Software search
.Dl BSD only
.Dl Use with Nm bktrctl
.Dl Needs kernel with Ql option BT848_GPIO_ACCESS
.Pp
.Em Guillemot MaxiRadio FM 2000 Radio Card
.Pq PCI
.Ql mr
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em AIMS Lab RadioTrack/RadioReveal RA300 FM Radio Card
.Pq ISA
.Ql rt
.Dl port 0x20c - rt1
.Dl port 0x30c - rt2
.Dl Volume - 0 .. 10
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em AIMS Lab RadioTrack II FM Radio Card
.Pq ISA
.Ql rtII
.Dl port 0x20c - rtII1
.Dl port 0x30c - rtII2
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em SoundForte Awesome 64R SF64-PCE2 Driver
.Pq PCI
.Ql sae
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware Search
.Pp
.Em SoundForte Legacy 128 SF16-FMD2 Driver
.Pq ISA
.Ql sf2d
.Dl port 0x284 - sf2d1
.Dl port 0x384 - sf2d2
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Search not implemented yet
.Dl Detection not implemented yet
.Pp
.Em SoundForte RadioX SF16-FMI Driver
.Pq ISAPNP
.Ql sfi
.Dl Also known as Creative ViBRA 16c PnP
.Dl port 0x284 - sfi1
.Dl port 0x384 - sfi2
.Dl Volume - 0 .. 1
.Dl Can set mono - no
.Dl Search not implemented yet
.Dl Detection not implemented yet
.Pp
.Em SoundForte RadioLink SF16-FMR FM Radio Card
.Pq ISA
.Ql sfr
.Dl port 0x284 - sfr1
.Dl port 0x384 - sfr2
.Dl Volume - 0 .. 15
.Dl Can set mono - no
.Dl Use with Nm fmrinit
.Dl Search not implemented yet
.Pp
.Em SoundForte RadioLink SF16-FMR2 FM Radio Card
.Pq ISA
.Ql sf2r
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em SoundForte RadioLink SF64-PCR FM Radio Card
.Pq PCI
.Ql sf4r
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em Spase PCRadio Card
.Pq ISA
.Ql sp
.Dl port 0x1b0 - sp1
.Dl port 0x1f0 - sp2
.Dl port 0x278 - sp3
.Dl port 0x378 - sp4
.Dl port 0x2f8 - sp5
.Dl port 0x3bc - sp6
.Dl Volume - 0 .. 63
.Dl Can set mono - yes 
.Dl Search not implemented yet
.Pp
.Em SoundForte Quad X-treme SF256-PCP-R Driver
.Pq PCI
.Ql sqx
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em SoundForte Theatre X-treme 5.1 SF256-PCS-R Driver
.Pq PCI
.Ql stx
.Dl Volume - 0 .. 1
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em Sound Vision 16 Gold Sound Card with FM Radio
.Pq ISA
.Ql svg
.Dl Volume - 0 .. 1
.Dl Can set mono - no
.Dl Software search
.Pp
.Em Trust FM Radio Card
.Pq ISA
.Ql tr
.Dl Volume - 0 .. 63
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em Terratec FM Radio Card
.Pq ISAPNP
.Ql tt
.Dl Volume - 0 .. 7
.Dl Can set mono - yes
.Dl Hardware search
.Pp
.Em Video4Linux Driver
.Ql v4l
.Dl Volume - 0 .. 10
.Dl Can set mono - yes
.Dl Software search
.Pp
.Em Zoltrix RadioPlus 108 FM Radio Card
.Pq ISA
.Ql zx
.Dl port 0x20c - zx1
.Dl port 0x30c - zx2
.Dl Volume - 0 .. 16
.Dl Can set mono - yes
.Dl Software search
.Sh EXAMPLES
.Bl -item
.It
Set frequency to 104.5 MHz:
.Bd -literal -offset indent
# fmio -f 104.5
.Ed
.It
Set frequency to 104.5 MHz and show signal state on this frequency:
.Bd -literal -offset indent
# fmio -f 104.5 -s
.Ed
.It
Set frequency to 104.5 MHz, hardware volume to 4:
.Bd -literal -offset indent
# fmio -v 4 -f 104.5
.Ed
.It
Set volume of both channels of line input to 9:
.Bd -literal -offset indent
# fmio -x line=9
.Ed
.It
For line1, increase volume of left channel by 2, set volume of right channel
to 10:
.Bd -literal -offset indent
# fmio -x line1=+2,10
.Ed
.It
For master output, increase volume of left channel by 2 and decrease volume
of right channel by 1:
.Bd -literal -offset indent
# fmio -X +2,-1
.Ed
.It
Decrease volume of both channels by 2:
.Bd -literal -offset indent
# fmio -X -2
.Ed
.It
Scan range [100.0...107.6] using driver
.Sq bktr :
.Bd -literal -offset indent
# fmio -d bktr -S -l 100.0 -h 107.6
.Ed
.It
Search station below 104.3 MHz using driver
.Sq sf2r :
.Bd -literal -offset indent
# fmio -d sf2r -W -104.3
.Ed
.It
Detect all available cards:
.Bd -literal -offset indent
# fmio -D
.Ed
.El
.Sh FILES
.Bl -tag -width /dev/mixer
.It Pa /dev/mixer
mixer audio device
.It Pa /dev/tuner
radio tuner device used by the
.Bx
bktr driver.
.It Pa /dev/radio
radio tuner device used by 
.Ql v4l
driver
.Pq Linux
and
.Ql br
driver
.Pq OpenBSD and NetBSD
.El
.Sh ENVIROMENT
The following environment variables affect the execution of
.Nm fmio :
.Bl -tag -width FMTUNER
.It Ev FMTUNER
The driver that should be used as default.
.It Ev RADIODEVICE
The radio tuner device
.Pq OpenBSD, NetBSD and Linux .
.It Ev MIXERDEVICE
The audio mixer device to use.
.El
.Sh BUGS
.Pp
Frequency and hardware volume should be absolute.
They cannot be increased or decreased relatively to the current value.
.Pp
The Gemtek PCI card implementation has some flaw.
Namely, when turning a card off, don't use option 
.Fl f .
.Pp
The SF16-FMR card has 16 volume levels.
Thus,
.Ic fmio -f 104.5
can be easily taken in as
.Ic fmio -v 0 .
Use some volume level when setting up a frequency.
Also, this card has to be initialized during bootup.
Launch
.Nm fmrinit
from some start script
.Pq most likely Pa /etc/rc.local .
.Pp
There's a flaw in the
.Bx
bktr driver - after closing
.Pa /dev/tuner
a card is turned off too.
Thus, you cannot use
.Nm
alone to manage bktr cards because
.Nm
opens
.Pa /dev/tuner ,
tunes a card to some frequency and closes
.Pa /dev/tuner
- and a card is turned off.
Since
.Nm
cannot hold
.Pa /dev/tuner
open all the time, use it with
.Nm bktrctl .
The same concerns Linux users.
Some cards supported by kernel are turned off after closing
.Pa /dev/radio
.Pq e.g. Tn USB tuners .
Use
.Nm
with
.Nm bktrctl .
.Pp
Due to the above limitation the
.Bx
bktr driver cannot work correctly if
frequency was not specified.
E.g.
.Ic fmio -d bktr -i
will not work as expected.
Correct use is
.Ic fmio -d bktr -f <some freq> -i .
.Sh AUTHOR
Vladimir Popov
.Aq jumbo@narod.ru


src/fmio.c
/*
 * Copyright (c) 2000 - 2002 Vladimir Popov <jumbo@narod.ru>.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
/*
 * $Id: fmio.c,v 1.71 2002/01/12 09:48:46 pva Exp $
 * core file for fmio utility
 */

#ifndef __DOS__
#include <err.h>
#endif

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef __DOS__
#include <stdint.h>
typedef uint32_t	u_int32_t;
typedef uint16_t	u_int16_t;
typedef uint8_t		u_int8_t;
#endif /* __DOS__ */

#ifdef __QNXNTO__
#include <inttypes.h>
typedef uint8_t u_int8_t;
typedef uint16_t u_int16_t;
typedef uint32_t u_int32_t;
#define NOMIXER
#endif /* __QNXNTO__ */

#include "radio.h"

#include "config.h"

/* Actions */
/* major */
#define NONE	0x0000
#define SCAN	0x0100
#define DETE	0x0200
#define SRCH	0x0400
/* minor */
#define MINOR	0x00FF
#define STAT	0x0001
#define TUNE	0x0002
#define VOLU	0x0004
#define MONO	0x0008
#define INFO	0x0010

char *pn = NULL;

void die(int);
void usage(void);
void invalid_driver_error(char *);
int gouser(void);
int goroot(void);

int
main(int argc, char **argv) {
	int optchar;
	int search = 0;
	char *drv = NULL;
	u_int16_t freq = DEF_FREQ;
	int volu = 0;
	u_int16_t action = NONE;
	u_int16_t lower = 0, higher = 0;
	u_int32_t cycle = 1;
#ifndef NOMIXER
	int mixer = 0;
	char *master_volume = NULL;
	char *other_volume = NULL;
#endif /* !NOMIXER */

	if (gouser() < 0)
		return -1;

	pn = strrchr(argv[0], '/');
	if (pn == NULL)
		pn = argv[0];
	else
		pn++;

	radio_init();

	/* 
	 * Call radio_drv_init() before usage(),
	 * or default driver will be: NULL, 0x0
	 */
	drv = getenv("FMTUNER");
	if (drv == NULL || *drv == '\0')
		drv = DEF_DRV;

	if (radio_drv_init(drv) == ERADIO_INVL)
		invalid_driver_error(drv);

	if (argc < 2)
		usage();

	/* Argh... options */
#ifndef NOMIXER
	while ((optchar = getopt(argc, argv, "c:Dd:f:h:il:mSsv:W:X:x:")) != -1) {
#else
	while ((optchar = getopt(argc, argv, "c:Dd:f:h:il:mSsv:W:X:x:")) != -1) {
#endif /* !NOMIXER */
		switch (optchar) {
		case 'c': /* number of probes for each scanned frequency */
			if ((cycle = strtol(optarg, (char **)NULL, 10)) == 0)
				cycle = 1;
			break;
		case 'D':
			action = DETE;
			break;
		case 'd':
			radio_drv_free();
			if (radio_drv_init(optarg) == ERADIO_INVL)
				invalid_driver_error(optarg);
			break;
		case 'f':
			freq = 100.0 * strtod(optarg, (char **)NULL);
			if (freq == 0)
				freq = DEF_FREQ;
			action |= TUNE;
			break;
		case 'h':
			higher = atof(optarg) * 100;
			break;
		case 'i':
			action |= INFO;
			break;
		case 'l':
			lower = atof(optarg) * 100;
			break;
		case 'm':
			action |= MONO;
			break;
		case 'S':
			action = SCAN;
			break;
		case 's':
			action |= STAT;
			break;
		case 'v':
			volu = strtoul(optarg, (char **)NULL, 10);
			action |= VOLU;
			break;
		case 'W':
			action = SRCH;
			search = atof(optarg) * 100;
			break;
#ifndef NOMIXER
		case 'X': /* set outputs.master */
			master_volume = optarg;
			break;
		case 'x':
			other_volume = optarg;
			break;
#endif /* !NOMIXER */
		default:
			usage();
		}
	}

	/* Minor actions have more priority */
	if (action & MINOR) action &= MINOR;

#if 0
	/* Drop privs for drivers that don't need root */
	if ((action & ~MINOR) != DETE)
		if (!radio_info_root())
			setuid(getuid());
#endif

	/* Enabling communication with the radio port */
	if (radio_info_root())
		if (goroot() < 0)
			die(1);

	if (radio_get_port() < 0)
		die(1);

	/* Test for card presense */
	if (radio_test_port() != 1) {
		fprintf(stderr, "%s: card not found: ", pn);
		radio_info_show(stderr, radio_info_name(), radio_info_port());
		radio_free_port();
		die(1);
	}

	if (radio_info_root())
		if (gouser() < 0)
			die(1);

	/*
	 * During time consuming actions fmio can get
	 * some signal. Define action for such emergency.
	 * Though it's more suitable to define these under SCAN or DETE,
	 * it won't hurt if they'll be here.
	 */
	signal(SIGINT , die);
#ifndef __DOS__
	signal(SIGHUP , die);
#endif
	signal(SIGTERM, die);

	switch (action & ~MINOR) {
	case NONE:
#ifndef NOMIXER
		mixer = radio_mixer_init() < 0 ? 0 : 1;
#endif /* !NOMIXER */
		if (radio_info_root())
			if (goroot() < 0)
				die(1);
		if (action & MONO)
			radio_set_mono();
		switch (radio_info_policy()) {
		case 0:
			if (action & TUNE)
				radio_set_volume(action & VOLU ? volu : 1);
			break;
		case 1:
			if (action & TUNE)
				radio_set_volume(action & VOLU ?
						volu : radio_info_maxvol());
			break;
		}
		if (action & TUNE)
			radio_set_freq(freq);
		if (action & VOLU)
			radio_set_volume(volu);
		if (action & STAT) {
			int st = radio_info_stereo();
			int si = radio_info_signal();
			if (st != ERADIO_INVL)
				printf("%s", st ? "stereo" : "mono");
			if (st != ERADIO_INVL && si != ERADIO_INVL)
				printf(" : ");
			if (si != ERADIO_INVL)
				printf("%s", si ? "signal" : "noise");
			if (st != ERADIO_INVL || si != ERADIO_INVL)
				printf("\n");
		}
		if (action & INFO) {
			u_int16_t f = radio_info_freq();
			int v = radio_info_volume();

			printf("Driver: ");
			radio_info_show(stdout, radio_info_name(), radio_info_port());
			if (f)
				printf("Frequency: %.2f MHz\n", (float) f / 100);
			if (v)
				printf("Volume: %u\n", v);
			v = radio_info_signal();
			if (v != ERADIO_INVL)
				printf("Signal: %s\n", v ? "on" : "off");
			v = radio_info_stereo();
			if (v != ERADIO_INVL)
				printf("Stereo: %s\n", v ? "on" : "off");
		}
		if (radio_info_root())
			gouser();
#ifndef NOMIXER
		if (mixer)
			if (other_volume)
				radio_mixer_update_volume(other_volume);

		/*
		 * Now outputs.master=0,0
		 * Update saved values
		 */
		if (mixer) {
			if (master_volume)
				radio_mixer_set_volume(master_volume);
			radio_mixer_cleanup();
		}
#endif /* !NOMIXER */
		break;
	case DETE:
		if (goroot() < 0)
			die(1);
		radio_detect();
		if (gouser() < 0)
			die(1);
		break;
	case SCAN:
		if (radio_info_root())
			if (goroot() < 0)
				die(1);
		radio_scan(lower, higher, cycle);
		if (radio_info_root())
			if (gouser() < 0)
				die(1);
		break;
	case SRCH:
#ifndef NOMIXER
		mixer = radio_mixer_init() < 0 ? 0 : 1;
#endif /* !NOMIXER */
		if (radio_info_root())
			if (goroot() < 0)
				die(1);
		if (search < 0)
			freq = radio_search(0, -1 * search);
		else
			freq = radio_search(1, search);
		if (radio_info_root())
			if (gouser() < 0)
				die(1);
		if (freq)
			printf("%s: %.2f MHz\n", pn, (float)freq / 100);
#ifndef NOMIXER
		if (mixer)
			radio_mixer_cleanup();
#endif /* !NOMIXER */
		break;
	default:
		break;
	}

	if (radio_info_root())
		if (goroot() < 0)
			die(1);
	radio_free_port();
	if (radio_info_root())
		gouser();
	radio_drv_free();
	radio_cleanup();

	return 0;
}

/*
 * Dump the usage string to screen
 */
void
usage(void) {
	const char usage_string[] =
#ifdef NOMIXER
		"Usage:  %s [-d driver] [-f frequency] [-i] [-m] [-s] [-v volume]\n"
#else
		"Usage:  %s [-d drv] [-f freq] [-i] [-m] [-s] [-v vol] [-X vol] [-x vol]\n"
#endif /* NOMIXER */
		"\t%s [-d driver] -S [-l begin] [-h end] [-c count]\n"
		"\t%s [-d driver] -W frequency\n"
		"\t%s -D - detect driver\n\n"

		"\t-f frequency in Mhz, -f 98.0 for example\n"
		"\t-i information\n"
		"\t-m mono\n"
		"\t-s stat\n"
		"\t-v volume, -v 0 set tuner off\n"
		"\t-S scan -l start frequency, -h end frequency\n"
		"\t-c number of probes for each scanned frequency\n"
		"\t-W search\n"
	;
	printf("%s version %s\n", pn, VERSION);
	printf("Default driver: ");
	radio_info_show(stdout, radio_info_name(), radio_info_port());
	printf(usage_string, pn, pn, pn, pn);

	die(0);
}

void
die(int sig) {
	radio_drv_free();
	radio_cleanup();
	exit(sig);
}

/*
 * Complain about invalid driver and init with default driver
 */
void
invalid_driver_error(char *invl_drv) {
#ifdef __DOS__
	printf("Invalid driver `%s', using default `%s'", invl_drv, DEF_DRV);
#else
	warnx("Invalid driver `%s', using default `%s'", invl_drv, DEF_DRV);
#endif
	radio_drv_free();
	radio_drv_init(DEF_DRV);
}

int
gouser(void) {
#ifndef __DOS__
	if (seteuid(getuid()) < 0) {
		warn("set effective user %d privs error", getuid());
		return -1;
	}
#endif
	return 0;
}

int
goroot(void) {
#ifndef __DOS__
	if (seteuid(0) < 0) {
		warn("set effective root privs error");
		return -1;
	}
#endif
	return 0;
}
